# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
! pip install pandas

^C


In [1]:
import pandas as pd

C:\Users\viver\AppData\Local\Temp\ipykernel_18224\4080736814.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
df = pd.read_csv("../data/supermarket_sales.csv")

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [8]:
df.head(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


In [6]:
df.tail(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [10]:
df.loc[615:619]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
615,509-29-3912,A,Yangon,Member,Female,Sports and travel,38.72,9,17.4240,365.9040,3/20/2019,12:24,Ewallet,348.48,4.761905,17.4240,4.2
616,327-40-9673,B,Mandalay,Member,Male,Sports and travel,72.60,6,21.7800,457.3800,1/13/2019,19:51,Cash,435.60,4.761905,21.7800,6.9
617,840-19-2096,C,Naypyitaw,Member,Male,Electronic accessories,87.91,5,21.9775,461.5275,3/14/2019,18:10,Ewallet,439.55,4.761905,21.9775,4.4
618,828-46-6863,A,Yangon,Member,Male,Food and beverages,98.53,6,29.5590,620.7390,1/23/2019,11:22,Credit card,591.18,4.761905,29.5590,4.0
619,641-96-3695,C,Naypyitaw,Member,Female,Fashion accessories,43.46,6,13.0380,273.7980,2/7/2019,17:55,Ewallet,260.76,4.761905,13.0380,8.5


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [14]:
df.dtypes

Invoice ID                     str
Branch                         str
City                           str
Customer type                  str
Gender                         str
Product line                   str
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                           str
Time                           str
Payment                        str
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [15]:
pd.DataFrame(df.dtypes).T

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,str,str,str,str,str,str,float64,int64,float64,float64,str,str,str,float64,float64,float64,float64


Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [19]:
#df.describe()
df_selected_cols = df[['gross income']]

In [20]:
df_selected_cols.describe()

,gross income
count,1000.000000
mean,15.379369
std,11.708825
min,0.508500
25%,5.924875
50%,12.088000
75%,22.445250
max,49.650000


In [21]:
with pd.option_context("display.precision", 2):
    display(df_selected_cols.describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99]))

,gross income
count,1000.00
mean,15.38
std,11.71
min,0.51
1%,0.91
25%,5.92
50%,12.09
75%,22.45
99%,45.25
max,49.65


75й персентиль колонки "gross income" = 22.45
Це нам показує, що 75% записів з нашого датафрейму мають gross income <= 22.45,
В той час як решта 25% мають цей показник в діапазоні (22.45 : 45.25] - тобто доходить аж до вдвічі більшого значення.


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [22]:
#Invoice ID
#City

df_selected_cols_2 = df[['Invoice ID', 'City']]

In [24]:
df.groupby('City')['Invoice ID'].count()


City
Mandalay     332
Naypyitaw    328
Yangon       340
Name: Invoice ID, dtype: int64

In [25]:
df['City'].value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

In [26]:
df.groupby('City')['Invoice ID'].nunique()

City
Mandalay     332
Naypyitaw    328
Yangon       340
Name: Invoice ID, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [3]:
df['City'].value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [5]:
df_city = df[["City", "gross income"]]
df_city

,City,gross income
0,Yangon,26.1415
1,Naypyitaw,3.8200
2,Yangon,16.2155
3,Yangon,23.2880
4,Yangon,30.2085
...,...,...
995,Naypyitaw,2.0175
996,Mandalay,48.6900
997,Yangon,1.5920
998,Yangon,3.2910


In [6]:
filter_yangon = df['City'] == 'Yangon'
filter_mandalay = df['City'] == 'Mandalay'

In [7]:
mean_yangon = df[filter_yangon]['gross income'].mean()
mean_mandalay = df[filter_mandalay]['gross income'].mean()

(float(mean_yangon), float(mean_mandalay))

(14.874001470588235, 15.232024096385542)

In [8]:
if mean_yangon > mean_mandalay:
    print("Cереднє значення gross income для міста Yangon > Mandalay")
else:
    print("Cереднє значення gross income для міста Yangon < Mandalay")

Cереднє значення gross income для міста Yangon < Mandalay




Отримане середнє значення gross income для міста Yangon < Mandalay.
Але як видно - різниця зовсім незначна. В чому полягають особливості кожного з цих магазинів ми можемо легко дослідити з наявних колонок в нашій таблиці. Кожна з них містить цінну інформацію для відповіді на це питання. 
Яку б колонку ми не взяли в роботу - дізнаємось щось нове про профіль роботи магазину. Зокрема:
__________________
Гіпотеза 1: Тип клієнтів (Customer type)
За колонкою Customer type можемо перевірити гіпотезу "Постійний клієнт чи випадковий покупець" (Member чи Normal) - маючи на думці, що постійний клієнт схильний мати більший середній чек, тож якщо магазин має більше Member клієнтів то і середній чек цього магазину має бути більшим.

Гіпотеза 2: Гендерні відмінності покупців (Gender)
За колонкою Gender можемо перевірити гіпотезу  що чоловіки і жінки мають різне наповнення своїх продуктових кошиків (Sports and travel проти Health and beauty) що впливає на суму покупки.

Гіпотеза 3: Асортимент (Product line)
Колонка Product line (Electronic accessories, Food and beverages, Fashion accessories, Sports and travel, Health and beauty, Home and lifestyle) породжує гіпотезу що магазини можуть мати різний набір товарів на продаж і саме в цьомуу може критись причина різниці в mean gross income  - бо товари одних категорій будуть дорожчі за інші (Food and beverages < Electronic accessories).

Unit price - ця колонка підсилить аналіз 2 попередніх гіпотез, оскільки покаже середню/мін/макс вартість товарів кожної категорії.

Гіпотеза 4: Кількість товарів як соціальний профіль (Quantity)
Колонка Quantity - відобразить для нас соціальний профіль жителів міста як покупців. Якщо тут проживають переважно великі родини - наповнення чеків скоріше більш з Food and beverages та у великих обсягах, а якщо місто для молоді, неодружених, студентів - чеки малі, покупки одиничні, не систематичні.

Гіпотеза 5: Вплив часу шопінгу на розмір чеку (Date, Time)
На колонках Date, Time може грунуватись така гіпотеза про вплив часу шопінгу на розмір чеку. Так покупки зроблені в обідню перерву чи цілеспрямований шопінг "скупитись на всі вихідні" - це дві величезні різниці.

Гіпотеза 6: Вплив типу оплати на розмір чеку (Payment)
І наостанок з колонкою Payment - припускаємо, що при оплаті Ewallet чи Credit card мають бути в середньому значно більші середні чеки аніж при оплаті 
_______________________

Висновки
Оскільки в середньому gross income між Yangon та Mandalay є незначним, шлях до підвищення прибутку полягає в розумінні профілю свого покупця. Слід оптимізувати під нього асортимент наявних товарів. Що саме потребують локальні покупці - нам покажує переважання типів клієнтів, платіжних методів та розподіл покупок по днях тижня та годинах. 


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [9]:
df['gross margin percentage'].nunique()

1

In [10]:
if df['gross margin percentage'].nunique() == 1:
    print("У всіх всіх інвойсах однакова маржа")
else:
    print("Маржа відрізняється для різних інвойсів")

У всіх всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [11]:
#'Product line' 

filter_yangon = df['City'] == 'Yangon'
filter_mandalay = df['City'] == 'Mandalay'

In [12]:
product_yangon = df[filter_yangon]['Product line'].value_counts(normalize=True)
# *100
print(product_yangon)

print()

product_mandalay = df[filter_mandalay]['Product line'].value_counts(normalize=True)
# *100
print(product_mandalay)

Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64

Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)